### ISSUE TITLE
Universal rendering

### ISSUE THREAD

**Issue by mxstbr** [react-boilerplate/react-boilerplate#174]
Adding server-side rendering to this boilerplate is probably the n1 request I get, and it does look quite useful.

Me personally, I haven't played with that at all, so I'm not yet sure if I want to include it by default or not. What do you think?

https://gist.github.com/rgrove/3e612aa366541845161c

**Comment by luandro** [comment]
Hey, I've done quite a lot of Universal rendering setup in the past year. You can check the approach I've been using [here](https://github.com/luandro/hapi-universal-redux), which is mainly inspired by [react-isomorphic-staterkit](https://github.com/RickWong/react-isomorphic-starterkit).

I guess these are the challenges for SSR:
- server-side live-reloading
- server-side pre-fetching with Redux

I've seen the server reloading problem being solved in a few ways:
- use [webpack-dev-middleware]() and [webpack-hot-middleware]() if in `development` as seen [here](https://github.com/jaredpalmer/react-production-starter), or
- use [concurrently](https://github.com/kimmobrunfeldt/concurrently) and [just-wait](https://github.com/download/just-wait) and different webpack config files, as seen [here](https://github.com/RickWong/react-isomorphic-starterkit)

The best api for the pre-fetching problem I've seen has been with [redial](https://github.com/markdalgleish/redial).

Those two options seem to solve all the problems discussed on the [gist](https://gist.github.com/rgrove/3e612aa366541845161c). Am I missing something, or are these to naive?

I'd be glad to give it a try, let me know what you think.

**Comment by mxstbr** [comment]
We're really close to the [v3 release](https://github.com/mxstbr/react-boilerplate/issues?q=is%3Aopen+is%3Aissue+milestone%3Av3.0) with just a few bugfixes left. I expect we officially release and launch it in (at the very latest) a month. This version has seen a major architectural overhaul and massive improvements all around.

I don't think introducing another overhaul for v3 is worth it, but server side rendering is something we definitely need to at least try, and probably release with v4. (?) I'd personally **love to see a first implementation**, just don't expect your PR to be merged immediately!

> Note: If you're going to try this, work on the **v3.0.0 branch** please!

**Comment by luandro** [comment]
I'm gonna try to start working on this after #211 settles.

**Comment by mxstbr** [comment]
As per #211, we should provide alternative implementations in other backend systems. @oliverturner might be able to do a Hapi one, but before we start we need a clear set of specs that all systems need to cover.

I don't have that much experience with SSR, so maybe somebody could shed some light on those? @luandro, what would your implementation cover?

**Comment by luandro** [comment]
I maintain a very [basic setup](https://github.com/luandro/hapi-universal-redux/blob/master/src/server.js) for a hapi server. I've implemented it in express as well, and there's very little difference.

It's a basic setup for:
- React and ReactDOM
- Redux 
- React-Router

It has setup for Radium, which won't be needed. And I think a something like [redial](https://github.com/markdalgleish/redial) would be nice to have for server-side data loading. There's probably a good way to fit it with Sagas.

**Comment by mxstbr** [comment]
What about this? https://github.com/keystonejs/react-easy-universal It seems quite easy to swap out express for something else.

**Comment by luandro** [comment]
It does reduce boilerplate, but then we would be depending on one more library to be regularly maintained. It doesn't take that much code to achieve this, and it would give us a lot more freedom.

I think we would be hiding a lot of simple functionality from the user. By taking a look at my example you'll see that what's basically going on is:
- React-Router matches the path in the server, and routes to the correct component.
- ReactDOMServer serves an html string according to the route at initial render.
- Redux initial state is passed to the client in some way.

**Comment by rikkert** [comment]
@luandro since express is used now, are you able to have a stab at this?
I would like to see your implementation within this boilerplate.

**Comment by luandro** [comment]
@rikkert sure, I'll try giving it a go.

**Comment by luandro** [comment]
Ok, first thing is how should we serve the html with the server if we depend on the [html-webpack-plugin](https://github.com/ampedandwired/html-webpack-plugin) to inject a bunch of stuff on our `index.html` file? Express needs to do a `res.send()`, with the whole initial html plus the React part rendered to string. Check [here](https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/helpers/Html.js) for an example of what the initial html looks like.

Is there a way to use the html-webpack-plugin to inject stuff into a js file containing the initial html?

**Comment by fubhy** [comment]
@luandro I started a project to demo a decoupled add based on GraphQL/Relay and Drupal. It is based on this boilerplate and uses server-side rendering with webpack using node target and also uses html-webpack-plugin. I already showed it to @mxstbr. You can check it out here: https://github.com/fubhy/drupal-decoupled-app

One thing that is still missing is hot-swapping the generated server-side code so that the server-side code is always up-to-date with the client side code. I am currently trying to sort that out.

Feel free to copy anything you find there and re-introduce it to this boilerplate. Unfortunately I currently don't have the time myself to submit a PR.

**Comment by jurekbarth** [comment]
@luandro do you have any updates for us? Just wondering, because i would love to see a nice solution 👍

**Comment by luandro** [comment]
Hey @jurekbarth, I'm been pretty busy these last weeks teaching a course. It'll be over tomorrow, and I'll have time to work on this. Sorry for the unexpected delay, didn't know courses demanded so much time 👍

**Comment by jurekbarth** [comment]
@luandro no problem if i can help with anything let me know!

**Comment by matthewp** [comment]
Hey, just wanted to chime in on prefetching. One thing you might want to check out is using Zones (like with can-zone: https://github.com/canjs/can-zone)

This would eliminate the need to prefetch.  I'm not super familiar with Redux but I think you could do something like this:

```js
import Zone from "can-zone";

... 

const initialState = { query: params }

new Zone().run(function(){
  const store = configureStore(initialState)

  return store;
}).then(function(data){
  const state = data.store.getState();

  const html = renderToString( ... )
});
```

can-zone automatically tracks asynchronous stuff, so it can all be contained within your reducers.

**Comment by steida** [comment]
It's safe. I am using it in my own dev stack https://github.com/este/este. Feel free to inspire.
The nice and unique thing among React universal data fetching is "double rendering" pattern. With Este we don't need a router for that, we can prefetch all components. It's demonstrated on Firebase, but the same pattern can be used for XHR or anything. https://github.com/este/este/blob/62226e3850b85a9b533c081c06ddc8f4c473c36b/src/common/lib/redux-firebase/queryFirebase.js#L186

**Comment by oyeanuj** [comment]
@mxstbr Congrats on v3 launch. Just checking to see if this is a high priority? Also, have you seen and have any thoughts on the universal rendering approach of https://github.com/erikras/react-redux-universal-hot-example ?

**Comment by jaredpalmer** [comment]
@mxstbr my boilerplate project handles SSR with async reducers and lazy routes: http://github.com/jaredpalmer/react-production-starter . The main challenge is hot loading. It becomes increasingly difficult because of react-router's `match` having to be run on the client, the need to isolate a Root component for HMR, and reducer injection.

This project is awesome by the way. Congrats.

**Comment by mxstbr** [comment]
@oyeanuj It's the highest priority at the moment, we understand how important this feature is for any serious large scale application!

@jaredpalmer thanks, we'll definitely take a look!

**Comment by mxstbr** [comment]
Possibly interesting: https://www.youtube.com/watch?v=PnpfGy7q96U by @aickin

**Comment by nobleach** [comment]
One thing I'd like to stress. Babel has made it blatantly clear (via a statement on its website) that transpiling on the fly is *NOT* meant for any real production scenario. I've seen this pretty much every "boiler plate". If there is any real hope of having a viable server-side rendering component to this project, let's skip that mistake and create an Express/Hapi server that runs in ES5 and a webpack job that transpiles the project. (Yes, Node 6 does support quite a bit of ES6/2015, but not the `import` statement - which means any real usage is dead in the water)

**Comment by rikkert** [comment]
Totally agree, the prod webpack needs another build with target: node. This
can also aid in fixing FOUC which is a nice added bonus.

**Comment by nobleach** [comment]
Exactly. I suggest switching out whatwg-fetch for isomorphic-fetch. This will add a fair amount of complexity, but we can certainly draw inspiration from other projects.

**Comment by skywalkerlucas** [comment]
Hi Author,

This project is excellent. It is carefully written and easy to understand, as "made with love". And I can see it is gaining much popularity.

BTW, As @oyeanuj  suggest, the SSR solution in  https://github.com/erikras/react-redux-universal-hot-example is quite mature and solid, and I guess you can consult that project in implementing SSR for this app.

Finally, as now this app is well organized, my suggestion is that, if you are to add SSR feature for this app ( which I also think is very beneficial ), it would be better if you do not change current structure a lot, because SSR is very obscure to many people, and if you break current structure ( which is very clear and easy to understand ), it will make this app harder to understand. So my suggestion is to make SSR feature parallel with current set-up, and people can choose.

This is a great project and thanks for your kindness of sharing!

**Comment by nobleach** [comment]
My apologies, but I completely disagree with the SSR solution presented in [https://github.com/erikras/react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example) That project is more of a demo that one could deploy to Heroku as a proof of concept. It is set up to transpile ES6 to ES5 on the fly. The babel project recommends against this approach. This would incur high memory usage and very bad performance. Luckily, SSR can be added to this project without moving anything else around.

**Comment by oyeanuj** [comment]
@nobleach Thanks for the performance context. Any project out there that you think does 'universal' well that one could take inspiration from?

**Comment by luandro** [comment]
What do you guys think about approach used in [https://github.com/RickWong/react-isomorphic-starterkit/](https://github.com/RickWong/react-isomorphic-starterkit/)?

**Comment by Hyra** [comment]
If it helps, I had a stab at a minimalistic SSR approach at https://github.com/Hyra/Antler

**Comment by oyeanuj** [comment]
And finally, one other reference point is the redfin react-server package. Admittedly, there is a scope is a little wider than this task but I think it has a lot of good things going that would be great to derive inspiration from - https://github.com/redfin/react-server

**Comment by steida** [comment]
A lot of starter kits are using plain `JSON.stringify`, which is unsafe. Serialize is recommended instead. 
https://github.com/este/este/blob/master/src/server/frontend/render.js#L45

One think I would love to see is CSP support by default. Someone suggested PR for that https://github.com/este/este/pull/731 Interesting stuff.

**Comment by oliverturner** [comment]
@nobleach not to be contrarian, but I'd love to see some data supporting the assertion that Babel server-side is slow/memory-intensive. When the server is running in production mode the code isn't transpiling on the fly: it's cached (as evidenced by the fact that changes to source code won't show up until the server is restarted).

We have a project about to go into production that uses Babel on the server and our finding is that response times and resource usage is well within the normal/expected range. We'll be doing load-testing presently, so I should be able to share some of the stats soon.

**Comment by aickin** [comment]
@oliverturner FWIW, that's my experience, too. 

Babel says nothing on the [Require Hook page](https://babeljs.io/docs/usage/require/), although there is a warning on their [setup page](https://babeljs.io/docs/setup/) if you select "Require Hook" that says:

> Not meant for production use
> The require hook is primarily recommended for simple cases.

I'm not sure what this is referring to, though; it might just mean that the require hook is a blunt tool that affects your app globally and therefore can be problematic. It's not obvious that they are talking about memory usage or perf.

**Comment by oliverturner** [comment]
@aickin Might be time to consult the Babel guys themselves... brb!

**Comment by luandro** [comment]
Just saw [https://github.com/airbnb/hypernova](https://github.com/airbnb/hypernova), which seems more mature than [https://github.com/keystonejs/react-easy-universal](https://github.com/keystonejs/react-easy-universal). It's an option if we wanna go the automatic way.

**Comment by nobleach** [comment]
I've been trying to get some hard data from my ops team. So far it's been "wow this thing is as heavy as starting up a JVM". Anecdotal evidence is not useful though.

**Comment by oliverturner** [comment]
"Anecdata" :)

**Comment by mxstbr** [comment]
> "wow this thing is as heavy as starting up a JVM". Anecdotal evidence is not useful though.

Let us know when there's a consensus :wink:

**Comment by justingreenberg** [comment]
i've been playing with hypernova over the past few days. it's extremely cool, but smells like a premature optimization for a boilerplate. since it's brand new (no abstractions) a "simple" implementation would require:
running a separate thread for renderer, 

i see hypernova + react-router having a baby, with something like redial to manage data fetching

**Comment by tzarger** [comment]
When is v4.0 planned to release? I am also interested in the Universal side of this awesome boilerplate.

**Comment by mxstbr** [comment]
No ETA, since we all work on it in our free time it'll be done when somebody gets around to do it! :+1:

**Comment by steida** [comment]
"Server side rendering is really trivial, but I will not push my stuff here." Some guy on internets :-D

**Comment by Stalso** [comment]
Hello. I am trying to implement server-side rendering in this boilerplate. But I am new in react and webpack (and in whole this node.js stuff) and can make many stupid mistakes. 

I have decided to make webpack  build of server side and then run already bundled server like it is implemented [here](https://github.com/choonkending/react-webpack-node) . I think this is the most suitable approach to implement ssr. 

The most difficult part for me, when I make server development build, is importing `import 'sanitize.css/sanitize.css';` in `app\containers\App\index.js`. I have discovered several solutions:

1. Use [webpack-isomorphic-tools](https://github.com/halt-hammerzeit/webpack-isomorphic-tools). But I do not like this approach. It has several limitations

2. Use  [isomorphic-style-loader](https://github.com/kriasoft/isomorphic-style-loader). I also do not like this approach. It forces to use their `insertCss()` functions.

3. Use [babel-plugin-webpack-loaders](https://github.com/istarkov/babel-plugin-webpack-loaders) . It works!!! But it is suitable only for babel projects. If I want to rewrite template with typescript it will not help me.

4. The most suitable way I have found - is using [universal-webpack](https://github.com/halt-hammerzeit/universal-webpack) . In fact this library helps to make server webpack build and run bundled code. But I can not make it working in the boilerplate. My issue is described [here](https://github.com/halt-hammerzeit/universal-webpack/issues/6) 

As I can not manage to setup [universal-webpack](https://github.com/halt-hammerzeit/universal-webpack) library, I have tried to make my own webpack config to transpile server side. And the big ptoblem for me is `import 'sanitize.css/sanitize.css';`  as I said. I have tried different loaders e.t.c But I am always getting the same error: 

    E:\sandbox\node\reactv10\node_modules\sanitize.css\sanitize.css:9 
    abbr[title] {
                    ^
    SyntaxError: Unexpected token {
        at Object.exports.runInThisContext (vm.js:53:16)
        at Module._compile (module.js:513:28)
        at Object.Module._extensions..js (module.js:550:10)
        at Module.load (module.js:458:32)
        at tryModuleLoad (module.js:417:12)
        at Function.Module._load (module.js:409:3)
        at Module.require (module.js:468:17)
        at require (internal/module.js:20:19)
        at eval (eval at <anonymous> (E:\sandbox\node\reactv10\build\server.js:4:8541),    <anonymous>:1:18)
        at Object.<anonymous> (E:\sandbox\node\reactv10\build\server.js:4:8541)

Keep struggling...

**Comment by andyhu** [comment]
What about https://github.com/kriasoft/react-starter-kit ? It seems to be so popular now

**Comment by Stalso** [comment]
@andyhu they use isomorphic style loader.  If you use it,  you need to inject their function calls everywhere

**Comment by philiplopez** [comment]
I haven't had a chance to work through using this myself, but I believe a universal solution for CSS Modules is https://github.com/css-modules/css-modules-require-hook/ - there is a [demo project](https://github.com/css-modules/css-modules-require-hook/tree/master/demo) as well. I believe this would avoid webpack-isomorphic-tools or universal-webpack for CSS at least (but not images, etc.).

There are a few other boilerplates/demos using the require-hook as per its [npm page](https://www.npmjs.com/package/css-modules-electron), e.g. https://github.com/StevenIseki/react-router-redux-example.

As I said, I haven't had a chance to tinker with CMRH or the above demos but my investigations to date indicated that it might be a solution that is compatible with react-boilerplate. I might get some time in the coming week to investigate further but otherwise I hope it helps others.

**Comment by suhaotian** [comment]
I think default is not. 
Because [react-starter-kit](https://github.com/kriasoft/react-starter-kit) did it.
And the charm for this `react-boilerplate` that it is  just a pure front-end boilerplate.

**Comment by Rampage89** [comment]
Hello All, Do you have any idea when the server side rendering can be implemented?

**Comment by mxstbr** [comment]
> when the server side rendering can be implemented

Now? 😁

**Comment by oliverturner** [comment]
Some thoughts:
- Not a fan of server-side Webpack: too much abstraction at the cost of debuggability
- I have a working SSR implementation with Redux, React Router and full PostCSS/CSS Module support: uses `babel-plugin-css-modules-transform` which builds on `require-hook`. It's awesome... but the integration is against Hapi, not Express.
- Despite the fact that this is a solved issue, my experience means that I'm not convinced that this project _should_ bundle Universal Rendering

**Comment by Rampage89** [comment]
Max :)  

Olivier, can you please elaborate why SSR should not be part of the project? Do you see it breaking the minimal working solution principle or you just don't see any benefit? 

To be honest, i am not experienced java developer. However, we are starting new react web project and it is up to me to identify, what architecture we are going to use. This template looks great for us. The only thing what we are missing is the SSR.

**Comment by tzarger** [comment]
@Rampage89 I agree, I believe moving this project to SSR is about the only thing missing. Many projects would need SSR in order to help with SEO requirements, etc. This project has soooo many great things, it would be tough to re-create this all in a separate project in order to have SSR, so I hope that it does indeed make it in. @mxstbr

**Comment by oliverturner** [comment]
Hey @Rampage89 (amazing handle btw) @tzarger
> Do you see it breaking the minimal working solution principle
That's exactly my objection

> or you just don't see any benefit?
The amount of value will depend on the kind of project you're building. SSR isn't universally useful...

> Many projects would need SSR in order to help with SEO requirements, etc.
This is true, but many - "applications" rather than content-focused "sites" - will not.

> This project has soooo many great features, it would be tough to re-create this all in a separate project in order to have SSR
It certainly does... but I'm not suggesting that you need a separate project for it because...

TL;DR
I'd rather provide solid documentation on how to implement SSR than bake it in to the project

**Comment by azampagl** [comment]
Just wanted to chime in before SSR/Universal (and #625) gets much further...

One **huge** benefit of this project is the build process providing static files. This means that any app built can be:

1) Placed on something like Amazon S3 (or whatever super fast static-file server) to serve flat files ridiculously fast. One would then be "forced" to completely separate out their API. Authentication could be achieved via JSONWebTokens/Auth0, if necessary.

2) This is a huge one that isn't mentioned often here... This project can be used in conjunction with Cordova (https://cordova.apache.org/) and your favorite CSS Framework to create awesome mobile apps using the latest web technologies. Because it is static files, it can be run in offline mode to boot! If SSR/Universal is the only option, then this will no longer be possible. @mxstbr Would docs on how to implement a mobile app using this bootstrap be of interest?

I'd agree with @oliverturner comments. I imagine the majority of developers will be developing apps, not sites. Not sure the benefits of SEO for one or two pages/slight speed ups is worth the complexity/confinement (see http://jamesknelson.com/universal-react-youre-doing-it-wrong/)

@datapimp, is it possible for your SSR implementation to support a static file build process as well as SSR so it's possible to get the best of both worlds?

**Comment by zanedev** [comment]
+1 Please keep universal rendering optional, static is the way to go for most of the use cases.

**Comment by jwinn** [comment]
Here's my thoughts/suggestions. Let's add a reference SSR implementation to the docs. Then--or in tandem, create and finish the CLI which allows for the option (at setup) to add SSR support (and any other optional features).

**Comment by tzarger** [comment]
@azampagl Think of an app like Zillow.com ... public site, built with React needing SEO. I am not talking about a 1 or 2 page site that needs SEO.

@jwinn I like the idea of a CLI option that adds SSR.

@oliverturner I see your points and could agree. Perhaps if not a CLI Option like @jwinn has suggested, at minimum some good documentation where all affected files would need to be modified, or perhaps a Gist.

**Comment by azampagl** [comment]
@tzarger Good point, completely agree.  But I imagine a large portion of the users have an app where most of the pages sit behind some type of authentication. I definitely think there needs to be an *option* for some type of SSR for SEO, even at a minimum a guide. I'd just hate to see the static build option go away in favor of a completely SSR/Universal solution.

**Comment by tzarger** [comment]
@azampagl I agree, there are definitely lots of sites that would not need SEO (primary example for SSR). I think a CLI option during setup would be the ideal scenario. Similar to the way ngrok was removed as default but yet you can still boot it up with npm run start:tunnel.

**Comment by oyeanuj** [comment]
To add my $0.02 to this as well - this boilerplate already has a lot of features which are most applicable or useful to more complex applications - offline, redux-saga, code-splitting are a few examples. As a similar best practice for applications, I think SSR is probably a healthy default to have (with an option obviously to be able to remove it). 

FWIW, all similar boilerplates have a SSR implementation, and having it here, would really propel this boilerplate's usage even more.

**Comment by tzarger** [comment]
@oyeanuj Yes, I agree many boilerplates do have SSR and yes I believe that would make this one more attractive too. Whether the default is SSR and then you remove or default is as is (no SSR), then add SSR, either way, as long as there is an option to get SSR whatever way makes sense.

It would be great if it was a setup option similar to generating a Component where it asks:
```
  ES6 Component
> Stateless
```

Where setup could ask:
```
> Client Side Rendering
  Server Side Rendering
```

then continue to run setup accordingly.

@oliverturner @jwinn @azampagl @datapimp

**Comment by gihrig** [comment]
I _really_ advocate for _optional_ SSR. 

I started my boilerplate/starter kit journey with [erikras/react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example) because it had "everything".

I eventually moved on because the complexity was overwhelming.

[Find your perfect React starter project](http://andrewhfarmer.com/starter-project/) lists 94 projects. This project is the only one that lists offline-first as a feature.

I would argue for quality over quantity. There seems to be more than enough quantity in the world. I would stay away from SSR just for the sake of making the project more visible.

Having said that, SSR that can be easily added or removed would be great, I have projects that would take advantage of it. I would prefer documentation over automation, because it helps me learn and put me in control.

**Comment by tzarger** [comment]
@gihrig I concur on an optional SSR configuration.

Where this project tends to be more complicated to add SSR is due to the use of redux-saga, which by the way, I really like, and having some great documentation would really help. With that, if you have great documentation, having some automation would be great so every time you start a new project that required SSR, you do not have to go through the steps. After all, if you did not want to use the automation so you could follow the documentation to learn, that could be your choice.

I agree that this boilerplate really stands apart and has some really great features as is.

**Comment by andyhu** [comment]
@tzarger How about replacing `redux-saga` with `redux-observable` + 'rx'? It looks very promising and interesting.
Check https://github.com/davezuko/react-redux-starter-kit/pull/828

**Comment by tzarger** [comment]
Hi @andyhu I have head RxJS is up and coming ... and some of the people I have heard using it swear by it. I would definitely be open to it.

However, I am not the one to make those kind of decisions, so I would pass that on to @mxstbr who is pretty strongly set on redux-saga.

**Comment by Stalso** [comment]
And I think,  that ssr is must have for react boilerplate. First of all,  ssr gives us not only SEO,  but also initial speed is important.  Also ssr provides for us possibility to make fast and good content based sites using react.  It is our job,  guys

**Comment by jaredpalmer** [comment]
@Stalso @gihrig @jwinn Solved these issues months ago....see [react-production-starter](https://github.com/jaredpalmer/react-production-starter). 

Features:
- React
- Redux, with async/split reducers (just like react-boilerplate)
- React-Router, with lazy routes (just like react-boilerplate)
- Webpack HMR
- Redial for data fetching, although it's trivial to roll your own Promise-based solution
- Aphrodite for styles, although also super easy to use PostCSS / whatever you want
- **...and SSR**

No magic.

**Comment by mxstbr** [comment]
@jaredpalmer I've deleted your comment. I don't mind references to other boilerplates/starter-kits at all, in fact I think it's a great place to get inspiration from. (I'm pretty sure others have taken inspiration from this one too!)

What you did though is plain advertising for your own, and that's a bit too much. Please refrain from doing that again, though we'd love to have your input about _how you solved the issues presented in this thread_.

**Comment by ankitduseja** [comment]
SSR is an important thing for some of sites which depend on SEO heavily. I agree that React was created with an aim to make single page apps better and easier, but the ecosystem surrounding react has led to many other great innovations (apart from react too) which are already a part of this repo and I am sure many people who need to build websites with SEO, would also want to use all the other good stuff too.

I was just thinking, does having a separate build command for ssr makes sense? Could that help people have a choice between the current implementation and the SSR one? If this could happen then people can choose if they want to deploy it on a static host like S3 or run their own node server.

@mxstbr I have been reading that a lot of discussion going on at several different threads in this repo regarding ssr. Can you please update regarding what have you decided or planning regarding SSR in this boilerplate after all this discussion and it would be great if you can you provide a timeline for the same?

**Comment by mxstbr** [comment]
> Can you please update regarding what have you decided or planning regarding SSR in this boilerplate after all this discussion and it would be great if you can you provide a timeline for the same?

I definitely know that SSR is the most important feature to get into this boilerplate. I realize it's important for many people, and I want to add it. We do want a great solution too, since a lot of people are going to be looking at it and using it!

However, I work on this boilerplate mostly in my free time, and have never done SSR before, so I can hardly provide a timeline I'm afraid! If you help implementing this feature I'm sure it'll be here faster :grin:

@datapimp kicked off some initial work at #625, but I'm not sure how long that'll take to get finished/if that's the right approach/…

**Comment by ankitduseja** [comment]
@mxstbr Thanks! Even I also haven't played around with SSR yet, but I'l help wherever I can.

**Comment by im-martijn** [comment]
Have you seen https://github.com/barbar/vortigern/blob/master/package.json ? They use SSR, I believe.

They use typescript too. Maybe it would be nice if that was something you could opt into in this boilerplate too?

**Comment by carlitux** [comment]
I was reading threads and I have some thoughts about this that maybe can help. I'd like to work on a pull request but I don't have knowledge on webpack and this will have many changes on structure. 

Main issue is: `require` css assets, images, fonts inside js code to work on nodejs. So I think https://www.npmjs.com/package/universal-webpack could be used to compile and get those requires working on node.

We can get SRR and actual setup with not issues and anyone can deploy as wanted/needed this because we can have this new directory structure when deploying:

```bash
./src  # here will be moved current app and server directories
./build  # files that will be served to browser
./server  # compiled version by webpack for nodejs of src
```

./build and ./server won't be same because plugins and loaders works different for node (I think it should do) because on server side webpack don't have to move image, fonts, or create the bundled css what server side rendering needs are hashes of image names, css class names and css filenames that matches to ./build.

./build should be deployable to server SPA as now do.
./server should be deployable and do SSR, anyone can deploy this if needs the feature, we have to use as base https://github.com/matthew-andrews/isomorphic-fetch and as @tzarger told it is need to see how use `redux-saga` on server side.

With this we can have minimal boilerplate on SRR and if a user needs can add middlewares, proxies to external servers.

Update docs as needed and maybe some suggest how to modify nginx setup to work as proxy to node server.

PD.- this doesn't need to add another command to compile server this has to be done on file changes.

**Comment by tzarger** [comment]
@carlitux Nice post, thanks. In regards to using Nginx to proxy node server, I do that already and can help there as it is pretty straightforward. Someone here was working on Docker support, which might be interesting as well so it be pre-configured for people in a docker file.

**Comment by fraadsbrandth** [comment]
Any news on this?
See it's scheduled in the v4.0, but it doesn't have any due date.

**Comment by mxstbr** [comment]
> See it's scheduled in the v4.0, but it doesn't have any due date.

If you look up 5 comments:

> I definitely know that SSR is the most important feature to get into this boilerplate. I realize it's important for many people, and I want to add it. We do want a great solution too, since a lot of people are going to be looking at it and using it!

> However, I work on this boilerplate mostly in my free time, and have never done SSR before, so I can hardly provide a timeline I'm afraid! If you help implementing this feature I'm sure it'll be here faster 😁

https://github.com/mxstbr/react-boilerplate/issues/174#issuecomment-231533210

**Comment by richardscarrott** [comment]
I've had success implementing serverside rendering on the [60fram.es react boilerplate](https://github.com/60frames/react-boilerplate) -- the reason I mention this is because I've split out the serverside hot reloading into [webpack-hot-server-middleware](https://github.com/60frames/webpack-hot-server-middleware) which, having noticed you also use both 'webpack-dev-middleware' and 'webpack-hot-middleware', could work well here.

**Comment by asood123** [comment]
We are getting closer to getting SSR: https://medium.com/open-collective/asking-for-a-new-feature-on-an-open-source-project-3305deba1356#.6ehgqgzhc

**Comment by scf4** [comment]
Note: Rendering user input from Redux can result in XSS with SSR. It's something often missed since React (client side) protects against it.

Make sure the SSR implementation for this boilerplate protects against this by sanitizing this stuff :)

**Comment by hadifarnoud** [comment]
we are in the process of choosing a boilerplate/framework for our react app. if SSR is close to release, please do tell (if possible with ETA). this is a fantastic boilerplate and it would be a shame not to use it. I'm in favour of this project rather than next.js

I'm not advanced react dev, so adding it myself is quite hard.

**Comment by scf4** [comment]
@hadifarnoud this is a pretty cool boilerplate if you're using GraphQL instead of REST: https://github.com/reactql

**Comment by hadifarnoud** [comment]
@scf4 not using GraphQL unfortunately

**Comment by oyeanuj** [comment]
@hadifarnoud If SSR isn't ready to be merged here, I'd recommend https://github.com/ctrlplusb/react-universally

**Comment by hpsharon** [comment]
Has anyone managed to add SSR to their boilerplate app and can assist?

**Comment by jabacchetta** [comment]
Wondering if [this repo](https://github.com/faceyspacey/react-universal-component) might help?

**Comment by blling** [comment]
If someone is still interesting in this, please reopen, will close this for now.

**Comment by jabacchetta** [comment]
For anyone who's in need of universal rendering, I'd highly recommend [Next.js](https://github.com/zeit/next.js/).
